index.js

```diff
@@ -114,23 +114,50 @@ module.exports.iterator = (emitter, event, options) => {
 		multiArgs: false
 	}, options);
 
+	let limit = -1;
+	if (options.limit !== undefined) {
+		if (
+			!(
+				options.limit >= 1 &&
+				(options.limit === Infinity || Number.isInteger(options.limit))
+			)
+		) {
+			throw new TypeError('The `limit` option should be an integer >=1');
```

Comment:
Promise-returning methods should *never* throw. They should always return a promise. So this should return the error in `Promise.reject()`. You also need to update the tests in https://github.com/sindresorhus/p-event/pull/16/files#diff-1dd241c4cd3fd1dd89c570cee98b79ddR276 to use `await t.throwsAsync()`.

Reply:
I'm following the same convention as `options.count`.

Reply:
BTW, I see `iterator` as a factory to create an async iterator. We should fail fast as long as we detects invalid options by throwing an error. The rejected promise should be reserved for any errors happening during the async iteration.

Reply:
Those are not the same cases. For `options.count`, it's inside a `new Promise` constructor, so the exception will be turned into a promise rejection. Here it will not.

Reply:
I see it now. Maybe we should throw an error synchronously for invalid `options.count` instead of deferring it to a rejected promise.

For `await pEvent...`, it catches both anyway.

Reply:
> Maybe we should throw an error synchronously for invalid `options.count` instead of deferring it to a rejected promise.

No, according to the Promise specification, Promise-returning APIs should *always* return promises, meaning they should reject and not synchronously throw. This is a common mistake though. https://twitter.com/jaffathecake/status/638659818996269056